### ISSUE TITLE
[webview_flutter] Add macOS support

### ISSUE THREAD

**Issue by stuartmorgan** [flutter/flutter#41725]
This would use WKWebView, as on iOS.

Blocked on #41722

## Current status (as of May 2024)

This continues to be blocked on full macOS platform view support, as noted above. Flutter 3.22 (current `stable`) has partial macOS platform view support, but does not yet support gestures. Gesture support has landed on `master`, however.

There is a [mostly complete PR to add macOS support to `flutter_webview_wkwebview`](https://github.com/flutter/packages/pull/6221) (see [this comment](https://github.com/flutter/flutter/issues/41725#issuecomment-1507065294) for known limitations), and the current expectation is that the PR will land after the next Flutter `stable` release cycle, since at that point gesture support should work on `stable`. It will not land before then, since having webviews not handle gesture support would cause significant confusion.

**Comment by Morteza-Rastgoo** [comment]
Is it supported now?

**Comment by rodydavis** [comment]
Looking to implement OAuth flow for MacOS, is this supported yet?

**Comment by stuartmorgan** [comment]
No, as indicated by the fact that the issue is still open.

The way to find out when this is implemented is to subscribe to the issue and look for an update closing it.

**Comment by ThinkDigitalSoftware** [comment]
> Looking to implement OAuth flow for MacOS, is this supported yet?

@rodydavis You can still do oath using the default browser and url_launcher. That's how I'm doing it

**Comment by rodydavis** [comment]
@ThinkDigitalSoftware I created a web view on the macOS side via a pop up and got it to work like that. But we definitely still need web views on macOS so we can execute js if needed or anything else.

**Comment by ThinkDigitalSoftware** [comment]
Yes, definitely. I'm waiting for this too

**Comment by MichelDiz** [comment]
Setting my Ward here to follow this.

**Comment by stuartmorgan** [comment]
Please use the Subscribe button to follow issues instead of posting me-too-type comments that email everyone.

**Comment by MichelDiz** [comment]
The Subscribe button has some kind of bug. I have tried, I wouldn't write if I hadn't tried. 

The button says that I am subscribed, however, in all issues that I enter it is written that I am subscribed... And when I press the button, it says "You’re not receiving notifications from this thread." which doesn't make any sense.

Screenshot from another issue that I'm not following. 

![Captura de Tela 2020-04-04 às 12 30 31](https://user-images.githubusercontent.com/6570955/78454771-86e83100-7670-11ea-9df6-552cbfa37547.png)

**Comment by sjapps** [comment]
Is there a good reason why this doesn't just work out of the box today?  Doesn't `WKWebView` have the same API for mac and iOS? Is the fear supporting other platforms?

**Comment by stuartmorgan** [comment]
> Is there a good reason why this doesn't just work out of the box today?

Please read issue description, which lists the blocking issue.

**Comment by vanelizarov** [comment]
As a temp solution, I wrote a small [plugin](https://pub.dev/packages/flutter_macos_webview) that lets you display native WKWebView 🙂

**Comment by jtoy** [comment]
@vanelizarov thx so much, it only works for mac? i think we should modify it to work across mac/ios

**Comment by vanelizarov** [comment]
@jtoy yeah, it's only for macOS

For iOS/Android I strongly suggest using `webview_flutter` cause it's more convenient way to integrate WebViews in applications

**Comment by ThinkDigitalSoftware** [comment]
@jtoy, as @vanelizarov mentioned, there's a plugin for that. His would need to be merged with webview_flutter in order to get it unified. @vanelizarov  is there a way you can just build it into the webview_flutter plugin and submit the PR?

**Comment by jtoy** [comment]
yes, I mean we should merge it into one ticket, I agree submitting a PR to the flitter official one might be best.

**Comment by vanelizarov** [comment]
@ThinkDigitalSoftware my plugin presents WebView inside NSViewController on top of the FlutterViewController. So, when PlatformViews support will be available for macOS, I'll update my package or merge it with `webview_flutter`

**Comment by aboutmydreams** [comment]
I hope it can be solved in 2021

**Comment by sheshnathiicmr** [comment]
@aboutmydreams just got your notification on mail and came here to say Happy New Year stranger 🥇

**Comment by subhash279** [comment]
@vanelizarov Thanks for the plugin.
Can we use this webView to handle google/fb login?
If so how do I get the token back from the webView once customer has successfully entered their credentials?

**Comment by stuartmorgan** [comment]
@subhash279 Please use the plugin's project to ask questions or report issues with that plugin. This issue is about the `webview_flutter plugin`.

**Comment by mentor33** [comment]
> @vanelizarov Thanks for the plugin.
> Can we use this webView to handle google/fb login?
> If so how do I get the token back from the webView once customer has successfully entered their credentials?

only ask questions related to plugin please

**Comment by sxmeer-ahmed** [comment]
Use case
I want to display a webview inside my Flutter desktop app.

Question
Are there any workaround I could do for now? Am I doing it wrong and I shouldn't even be using the webview?

flutter: ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
flutter: The following UnsupportedError was thrown building WebView(dirty, state: _WebViewState#a4635):
flutter: Unsupported operation: Trying to use the default webview implementation for TargetPlatform.macOS but
flutter: there isn't a default one
flutter:
flutter: The relevant error-causing widget was:
flutter:   WebView file:///Users/vince/dartsidedev/presentations/2019-12-11/presentation/lib/main.dart:82:14
flutter:

**Comment by sxmeer-ahmed** [comment]
url_launcher & WebView not work

flutter: Another exception was thrown: Unsupported operation: Trying to use the default webview implementation for TargetPlatform.fuchsia but there isn't a default one

**Comment by weifengsonos** [comment]
> As a temp solution, I wrote a small [plugin](https://pub.dev/packages/flutter_macos_webview) that lets you display native WKWebView 🙂

**Comment by weifengsonos** [comment]
> As a temp solution, I wrote a small [plugin](https://pub.dev/packages/flutter_macos_webview) that lets you display native WKWebView 🙂

Hi @vanelizarov,

I am getting the error below when attempting to install your plugin (flutter_macos_webview: ^0.0.4) on my Mac (Catalina v10.15.7).  Any idea how to fix it?  What is the "higher minimum deployment target"?

Launching lib/main.dart on macOS in debug mode...
Running pod install...
CocoaPods' output:
↳
      Preparing

    Analyzing dependencies

    Inspecting targets to integrate
      Using `ARCHS` setting to build architectures of target `Pods-Runner`: (``)

    Fetching external sources
    -> Fetching podspec for `FlutterMacOS` from `Flutter/ephemeral`
    -> Fetching podspec for `flutter_macos_webview` from `Flutter/ephemeral/.symlinks/plugins/flutter_macos_webview/macos`

    Resolving dependencies of `Podfile`
      CDN: trunk Relative path: CocoaPods-version.yml exists! Returning local because checking is only perfomed in repo update
    [!] CocoaPods could not find compatible versions for pod "flutter_macos_webview":
      In Podfile:
        flutter_macos_webview (from `Flutter/ephemeral/.symlinks/plugins/flutter_macos_webview/macos`)

    Specs satisfying the `flutter_macos_webview (from `Flutter/ephemeral/.symlinks/plugins/flutter_macos_webview/macos`)` dependency were found, but they required a higher minimum deployment target.

**Comment by marcorm** [comment]
> > Looking to implement OAuth flow for MacOS, is this supported yet?
> 
> @rodydavis You can still do oath using the default browser and url_launcher. That's how I'm doing it

Hi @rodydavis , after opening the browser, are you "going back" with deeplinking? In this case how dod you handle it?
On Mobile i'm doing this with uni_links, but it doesn't support macos.

**Comment by stuartmorgan** [comment]
Please see https://github.com/flutter/flutter/wiki/Issue-hygiene#comments-providing-workarounds

This is not the right place for detailed discussions of workarounds for specific use cases.

**Comment by cbenhagen** [comment]
Thanks to https://github.com/flutter/engine/pull/30853 this should now be possible to implement.

**Comment by esDotDev** [comment]
Can the team provide any insight on the remaining level of effort here, and if we can hope to expect this in the short-medium term?

**Comment by stuartmorgan** [comment]
This is still blocked on https://github.com/flutter/flutter/issues/41722. The PR above was incremental work, not a complete implementation. https://github.com/flutter/flutter/issues/41722 would be the right place to ask about remaining effort there.

**Comment by strongwill01** [comment]
Continuous attention

**Comment by akaboshinit** [comment]
>@vanelizarov
>As a temp solution, I wrote a small [plugin](https://pub.dev/packages/flutter_macos_webview) that lets you display native WKWebView 🙂
https://github.com/flutter/flutter/issues/41725#issuecomment-732445234

This package was great!

But it was not the behavior I was expecting.
I wanted to use a plugin that would display the webview inline within the app.
macOs version of something like this ↓
[flutter_inappwebview](https://pub.dev/packages/flutter_inappwebview)
[flutter_inappwebview#sample](https://inappwebview.dev/docs/in-app-webview/basic-usage/)

I think everyone gathered here is expecting
I've created a package that displays an inline webview! ↓
[flutter_inline_webview_macos](https://pub.dev/packages/flutter_inline_webview_macos)
[flutter_inline_webview_macos#example](https://pub.dev/packages/flutter_inline_webview_macos/example)
[flutter_inline_webview_macos#github](https://github.com/akaboshinit/flutter_inline_webview_macos/)

But I'm not familiar with swift, so it still doesn't work as expected!
I need your help! I need you to contribute!

**Comment by stuartmorgan** [comment]
> I wanted to use a plugin that would display the webview inline within the app.

It is not yet possible to make such a plugin that works correctly. See my comments above about the blocking issue, which continue to apply.

**Comment by gslender** [comment]
Wondering what update/progress is happening on this issue?

Focus in Flutter seems to be on things like a new rendering engine (why) and yet we still have inconsistent support across desktop compared to mobile. Flutter was supposed to promise a single framework for developing cross platform, but this type of issue (and the underlying lack of PlatformView across Flutter desktop).

As such, is there anything we can do to refocus the effort to start addressing this??

**Comment by gslender** [comment]
Note that it looks like PlatformView is starting to show signs of being unblocked too ??

flutter_view examples exist for both macOS and Windows 

https://github.com/flutter/flutter/tree/master/examples/flutter_view/

**Comment by stuartmorgan** [comment]
That example is unrelated to platform views. Also, this isn't the place to discuss platform views themselves; please follow the relevant issues for updates on that.

**Comment by stuartmorgan** [comment]
I have an initial proof of concept working at https://github.com/stuartmorgan/packages/tree/webview-macos. (It's still the case that platform view support for macOS itself isn't done yet, as noted above, but I wanted to find any other issues in parallel.)

The list of issues I encountered just getting an initial load-and-render working:
- [ ] The generated Pigeon Obj-C code doesn't compile on macOS due to the `import`; we would need to either fix https://github.com/flutter/flutter/issues/124724 or migrate [at least the very top layer of `webview_flutter` to Swift](https://docs.google.com/document/d/1XsaulkJA6_ZSpM7chkQLhQY25sQqhEMiqHMYzw8H85o/edit?resourcekey=0-_cUjF1c0iBvRLKfV-3gK5A#heading=h.reauq7bq8vak).
- [ ] https://github.com/flutter/flutter/issues/124723 prevents correct platform view initialization; the PoC uses a very hacky workaround that would fail in all but the simplest cases.
- [ ] The Dart mirror of `WKWebView` inherits from `UIView` which isn't correct for macOS. I've punted on that in the PoC since it wasn't a blocker for getting things compiling, but we will need to add an `NSView` and rework the class hierarchy (or we'll need to decide to slightly back away from exact mirroring there, and have a more abstract view that can be implemented with either one).
- [ ] Nice to haves:
    - [ ] `WKWebView` doesn't expose a `scrollView` on macOS, so the code for getting and setting scroll position doesn't work. In the PoC they throw `UnimplementedError`; we should look into implementing those with JS calls on macOS instead for consistency of API support.
    - [ ] The lack of support for `publish` creates a little bit of otherwise-unnecessary forking; https://github.com/flutter/flutter/issues/124721 would avoid the need for that.
    - [ ] Loading a page from assets won't work since I had to stub that out due to https://github.com/flutter/flutter/issues/47681 (We could ship an initial version without this, but would need to have it throw an `UnimplementedError` rather than just mysteriously not working as with the current PoC).

**Comment by cupidchan** [comment]
@stuartmorgan do you think it's now at a good point to release?

**Comment by BullsEye34** [comment]
@stuartmorgan , @cbracken I came across [this plugin](https://github.com/hlwhl/webview_cef) and [this plugin](https://github.com/jnschulze/flutter-webview-windows) which renders Chromium Embedded Framework and WebView2 quite well. I noticed that it uses a Texture instead of a PlatformView. So gestures work, scrolling works, everything seems really cool.

Can we not use the same logic to show WebViews for desktop in that way?

**Comment by stuartmorgan** [comment]
> Can we not use the same logic to show WebViews for desktop in that way?

I'm not sure what you are asking, those plugins *are* showing webviews for desktop.

If you mean, could `webview_flutter` do exactly what `webview_cef` is doing, then in theory, sure, but that's not the goal of this plugin, nor would having two copies of the same plugin be useful for the ecosystem. Anyone happy with the tradeoffs inherent to `webview_cef` can certainly choose to use it. `webview_flutter` for macOS will wrap the system-provided webview, displayed via platform views, just as it does on iOS and Android.

**Comment by BullsEye34** [comment]
@stuartmorgan , what I meant was that right now, PlatformViews are not supported on Desktop, and if what I'm saying is true, we need PlatformView to work in order to show native WebView on the Flutter Application.

The approach what the 2 mentioned plugins are taking are using Texture and Listener. CEF shows a Chromium WebView , I agree, but what if, we could show the native WebView using Texture and Listener instead of PlatformView?

**Comment by stuartmorgan** [comment]
> but what if, we could show the native WebView using Texture and Listener instead of PlatformView?

If someone wants to try to develop a WKWebView-based plugin based on pixel scraping and synthetic event creation, they can certainly do so. There are some fundamental issues that I would expect to range from difficult to impossible to resolve, but that doesn't mean it couldn't be potentially be useful for some use cases.

**Comment by BullsEye34** [comment]
> > but what if, we could show the native WebView using Texture and Listener instead of PlatformView?
> 
> If someone wants to try to develop a WKWebView-based plugin based on pixel scraping and synthetic event creation, they can certainly do so. There are some fundamental issues that I would expect to range from difficult to impossible to resolve, but that doesn't mean it couldn't be potentially be useful for some use cases.

Sounds really interesting and challenging. I wonder the complexity of it. I feel the visual part can be handled like that, and the logical part like href and stuff be passed on to the dart side in like a Stream or something for dart to handle it further on. 
I agree it's not an alternative to the official WebView plugin, but for what we have currently, I think that should do the trick?

**Comment by xiayuanquan** [comment]
MacOS-compatible webView, you can refer to this, I wrote a case, it should be satisfied。 

pub.dev: https://pub.dev/packages/macos_webview_kit

github: https://github.com/xiayuanquan/macos_webview_kit

**Comment by stuartmorgan** [comment]
The [proof of concept branch](https://github.com/flutter/flutter/issues/41725#issuecomment-1507065294) has been updated for the latest master; see the checklist there for the current state.

**Comment by nomo-app** [comment]
@stuartmorgan how is it going with the proof of concept WebView for macOS?
Would it be possible to just merge it into master even if some edge-cases are not working?

**Comment by ekuleshov** [comment]
> @stuartmorgan how is it going with the proof of concept WebView for macOS? Would it be possible to just merge it into master even if some edge-cases are not working?

Not sure if you need beta, main or a master Flutter channel for this to work, but you could use webview_flutter plugin directly from the poc git repo. E.g. declare it your pubspec like this:

```
dependency_overrides:
  webview_flutter_platform_interface:
    git:
      url: https://github.com/stuartmorgan/packages
      path: packages/webview_flutter/webview_flutter_platform_interface
      ref: webview-macos
  webview_flutter:
    git:
      url: https://github.com/stuartmorgan/packages
      path: packages/webview_flutter/webview_flutter
      ref: webview-macos
  webview_flutter_wkwebview:
    git:
      url: https://github.com/stuartmorgan/packages
      path: packages/webview_flutter/webview_flutter_wkwebview
      ref: webview-macos
  webview_flutter_android:
    git:
      url: https://github.com/stuartmorgan/packages
      path: packages/webview_flutter/webview_flutter_android
      ref: webview-macos
  webview_flutter_web:
    git:
      url: https://github.com/stuartmorgan/packages
      path: packages/webview_flutter/webview_flutter_web
      ref: webview-macos
```

Though I'm getting a compilation error on macos:

```
/Users/../.pub-cache/git/packages-4c1ff634b39b6dbcefca350d238ba16a2063c03c/packages/webview_flutter/webview_flutter_wkwebview/darwin/Classes/FWFWebViewFlutterWKWebViewExternalAPI.m:12:39: error: no known instance method for selector 'valuePublishedByPlugin:'
      (FWFInstanceManager *)[registry valuePublishedByPlugin:@"FLTWebViewFlutterPlugin"];
                                      ^~~~~~~~~~~~~~~~~~~~~~
```

Build for iOS is also failing:

```
Error (Xcode): ../../../.pub-cache/git/packages-4c1ff634b39b6dbcefca350d238ba16a2063c03c/packages/webview_flutter/webview_flutter_wkwebview/lib/src/webkit_webview_controller.dart:702:14: Error: The method 'AppKitView' isn't defined for the class 'WebKitWebViewWidget'.
```

**Comment by stuartmorgan** [comment]
> E.g. declare it your pubspec like this: [...]

Only `webview_flutter_wkwebview` is changed on that branch, so it's not clear to me why you are overriding all of the other packages.

**Comment by ekuleshov** [comment]
@stuartmorgan admittedly I wasn't sure what changed and also the `webview_flutter` plugin declares dependencies on the published modules (e.g. ios and webview_flutter_platform_interface), so the `pub get` was failing without some additional overrides (and maybe this branch is also behind of main for some changes).

Anyhow, any idea what is up with the build errors I'm getting in your poc?

**Comment by RakshithSarja** [comment]
Hello @stuartmorgan could you please update when can we expect the full webview support for both MAC and Windows? It would be helpful if you can share some rough time line on this as we are planning for a new project using webview in MAC and Windows.

**Comment by stuartmorgan** [comment]
> could you please update when can we expect the full webview support for both MAC and Windows?

This issue is only about macOS, not Windows.

I currently expect `webview_flutter` macOS support to land when full macOS platform view support reaches the `stable` channel.

**Comment by bahricanyesil** [comment]
Is there any update or any date estimation for this support? @stuartmorgan Also if possible any workaround or alternative ways?

**Comment by letrungdo** [comment]
@stuartmorgan webview_flutter_wkwebview macOS is nice, but the trackpad scrolling feature is not working yet

**Comment by cbenhagen** [comment]
@letrungdo you currently need to use the Flutter main/master channel for gestures to work

**Comment by junedr375** [comment]
Is there any update on this ?

**Comment by amorenew** [comment]
@stuartmorgan 
I tried your branch and it works now with Flutter 3.22.2 stable except I disabled setBackgroundColor as it is not supported on macOS
Thanks a lot for your effort

dependency_overrides:
  webview_flutter_wkwebview:
    git:
      url: https://github.com/stuartmorgan/packages
      path: packages/webview_flutter/webview_flutter_wkwebview
      ref: webview-macos

**Comment by ekuleshov** [comment]
@stuartmorgan is a common/good practice to add `FlutterMacOS` pod dependency into the darwin modules shared between ios and macos?

**Comment by stuartmorgan** [comment]
@ekuleshov Sorry, I'm not following the question. I don't know what you are referring to when you say "the darwin modules", or what you mean by "good practice" (or how this relates to the webview_flutter plugin).

**Comment by ekuleshov** [comment]
> @ekuleshov Sorry, I'm not following the question. I don't know what you are referring to when you say "the darwin modules", or what you mean by "good practice" (or how this relates to the webview_flutter plugin).

Here is the diff for `ios/Podfile.lock` after adding a `dependency_overrides` for `webview_flutter_wkwebview` from your `webview-macos` branch. Some other plugins also have `FlutterMacOS` dependency on iOS and some don't.

<img width="813" alt="image" src="https://github.com/user-attachments/assets/d9c8edba-dd01-4923-9722-d2a5953d4078">

**Comment by stuartmorgan** [comment]
I'm not sure what the question is then. Are you asking if it's good practice for a plugin that supports macOS to have a dependency on FlutterMacOS for its macOS? If so yes; if it didn't the plugin wouldn't build.

**Comment by ekuleshov** [comment]
> I'm not sure what the question is then. Are you asking if it's good practice for a plugin that supports macOS to have a dependency on FlutterMacOS for its macOS? If so yes; if it didn't the plugin wouldn't build.

I get the dependency in the macos's Podfile.lock, but that one was added in the ios's Podfile.lock

**Comment by stuartmorgan** [comment]
`Podfile.lock` contains the entire transitive dependency tree off all podspecs that a project depends on; that's how CocoaPods lock files work.

**Comment by ekuleshov** [comment]
> `Podfile.lock` contains the entire transitive dependency tree off all podspecs that a project depends on; that's how CocoaPods lock files work.

Right, but why the `FlutterMacOS` being added to iOS?

**Comment by stuartmorgan** [comment]
Because the `FlutterMacOS` pod is dependency of the `webview_flutter_wkwebview` pod.

Discussions of the core structure of Flutter plugins aren't on-topic for this issue; Discord, Stack Overflow, or some other more general forum would be the right place for this.

**Comment by amorenew** [comment]
I wonder where to track gesture support for macOS platform views, I searched but I couldn’t find the related issue
From the conversation here and the related PR gesture support is the blocker for it that why I want to subscribe to the gesture issue

**Comment by amorenew** [comment]
stuartmorgan I tested your branch, the mouse and scrolling working as expected on Flutter 3.24.0

**Comment by stuartmorgan** [comment]
Filed follow-ups for remaining functionality:
- https://github.com/flutter/flutter/issues/153773
- https://github.com/flutter/flutter/issues/153774
- https://github.com/flutter/flutter/issues/153775

**Comment by BullsEye34** [comment]
Hi @stuartmorgan , @cbracken and everyone involved Amazing work!

I noticed that this issue is closed, but the OP is outdated.
I think it would be wise to update the same for better information.

kudos!
